temba/reports/models.py

```diff
@@ -0,0 +1,153 @@
+    flows_skipped = models.PositiveIntegerField(default=0)
+    flows_processed = models.PositiveIntegerField(default=0)
+
+    @classmethod
+    def get_last_collection_process_status(cls, org):
+        last_dc: DataCollectionProcess = (
+            cls.objects.filter(Q(related_org=org) | Q(start_type=cls.TYPE_AUTO)).order_by("-started_on").first()
+        )
+        if last_dc:
+            data_status = {
+                "lastUpdated": str(last_dc.started_on),
+                "completed": bool(last_dc.completed_on),
+                "progress": (
+                    (last_dc.flows_skipped + last_dc.flows_processed) / last_dc.flows_total
+                    if last_dc.flows_total
+                    else 1
+                ),
+            }
+        else:
+            data_status = {"lastUpdated": None, "completed": True, "progress": 1}
+        return data_status
+
+
+class CollectedFlowResultsData(models.Model):
+    """
+    A model to store flow results data that is required for analytics reports, in more practical way.
+    {
+        result_key: {
+            # ids of all contacts answered with that category
+            category_key: [1, 4, 15],
+            another_category_key: [3, 7]
+        }
+    }
+    """
+
+    flow = models.OneToOneField("flows.Flow", on_delete=models.PROTECT, related_name="aggregated_results")
+    data = JSONField(default=dict)
+    last_updated = models.DateTimeField(auto_now=True)
+
+    @classmethod
+    def collect_results_data(cls, flow: Flow):
+        flow_results = {result["key"]: result["categories"] for result in flow.metadata["results"]}
+        result_keys = sorted(list(flow_results.keys()))
+        is_data_exists = cls.objects.filter(flow_id=flow.id).exists()
+        is_data_already_updated = is_data_exists and not flow.runs.filter(
+            modified_on__gt=flow.aggregated_results.last_updated
+        )
+        if is_data_already_updated or not result_keys:
+            return
+
+        # select runs data and build the aggregated data to save
+        def category_extractor(field):
+            return "results::jsonb #> '{%s, category}' as %s" % (field, field)
+
+        query = (
+            f"SELECT {', '.join(map(category_extractor, result_keys))}, array_agg(contact_id) as contact_id "
+            f"FROM flows_flowrun WHERE flow_id = {flow.id}"
+            f"GROUP BY {', '.join(result_keys)}"
+        )
+
+        final_data = {key: {category: set() for category in categories} for key, categories in flow_results.items()}
+        with connection.cursor() as cursor:
+            cursor.execute(query)
+            result_answers = cursor.fetchall()
+            for idx, result_key in enumerate(result_keys):
+                for record in result_answers:
+                    category_key = record[idx]
+                    if category_key:
+                        final_data[result_key][category_key].update(record[-1])
```

Comment:
I've sent it on Slack but just adding here as well.

![Screen Shot 2021-08-10 at 18 49 42](https://user-images.githubusercontent.com/5198870/129117058-a8aec002-2cd8-4292-a53b-b80341950dbc.png)

Reply:
This part of the code was refactored.